File: Search_in_a_Sorted_Array_of_Unknown_Size.py
# Version 1
# Time complexity : O(logn)
# Space Complexity : O(1)
# Yes it ran successfully on leetcode
# Any problems faced? Nothing major



# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        max_value = (2**31)-1
        i = 0
        j = 10000
        while i <= j:
            k = (i+j)//2
            val = reader.get(k) 
            
            if val == target:
                return k
            elif val == max_value:
                j = k-1
            elif target < val:
                j = k-1
            else:
                i = k +1
        return -1
            

File: SearchInASortedArrayOfUnknownSize.cs
/*
 * Time Complexity: 0(log(m))
 * Space Complexity: 0(1)
 */
namespace LeetCodeSubmission.BinarySearch1;

/**
 * // This is ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *     public int Get(int index) {}
 * }
 */

public class SearchInASortedArrayOfUnknownSize
{
    public int Search(ArrayReader reader, int target) {
        
        int low = 0;
        int high = 1;
        
        int idx = -1;
        
        while(low <= high){
            int mid = (high-low)/2 + low;
            
            if(reader.get(mid) == target){
                idx = mid;
                break;
            }
            
            else if(reader.get(mid) < target){
                
                low = mid+1;
                
                if(low == high){
                    high = 2*high;
                }
            }
            
            else if(reader.get(mid) > target){
                high = mid-1;
            }
            
        }
        
        return idx;
    }
}